### ISSUE TITLE
Bring Combiner back?

### ISSUE THREAD

**Issue by manuelma** [Spine-project/Spine-Toolbox#1622]
We used to have the Combiner project item that would merge DBs from source Data Stores into destination Data Stores. We dropped it in favor of doing the merging when executing the "target" Data Store, but now I believe that was a mistake.

For example, consider the following workflow:

![Screenshot from 2021-12-09 18-39-03](https://user-images.githubusercontent.com/11472756/145447541-bdf6abea-246d-4ed3-adbe-986950896f7a.png)

The jump from Run SpineOpt DA into CAES_DAinput modifies the latter to roll the model forward. But when CAES_DAinput executes in the next iteration of the loop, it imports back the data from the CAES_DA_diff and CAES_common which overwrites the work of the jump!

That wouldn't happen if we had a Combiner doing the merging.

And in general, executing Data Stores doesn't feel good as @jkiviluo was saying in another issue. Moreover, I feel that in general is the Item going *into* the Data Store who writes to it, not the Data Store itself (think of Importer, Tool). But in Data Store execution, is the DS who writes to itself which thus feels like an exception :(

What do you think? Should we resucitate the good old Combiner? And then we can think it's a Christmas miracle...

**Comment by soininen** [comment]
Why is the jump link in the description's figure connected to CAES_DAInput? If you're going to loop the Tool only, why not connect the jump back to Run SpineOpt DA?

**Comment by soininen** [comment]
There is a way to get around with it without reintroducing Combiner: having a second data store pointing to the same database as CAES_DAInput as Tool's second successor (sorry about the legacy link colors):

![loop-example](https://user-images.githubusercontent.com/19147159/145530656-e4f60846-27d2-4c8c-994d-61ba9edd0315.png)

For me this is would be the more logical solution, but I know I am probably the only one here who thinks so :) In the end I guess it is OK if we end up resurrecting Combiner.

In any case, I feel these discussions about Data stores show that there is something very fundamental wrong with the concept.

**Comment by manuelma** [comment]
That's a nice solution @soininen !

However if we accept this as the recommended solution, we would have to duplicate Data Stores every time the workflow needs to pass by them a second time.

> these discussions about Data stores show that there is something very fundamental wrong with the concept

I think the issue is exactly that sometimes we execute them, and sometimes we don't. The Combiner would bring more consistency I believe. But I don't know if you any strong objections?

**Comment by soininen** [comment]
>  we would have to duplicate Data Stores

I've been advocating duplicate Data Stores (having multiple Stores point to the same database) here and there but it seems to be a difficult concept to grasp to lots of users. This makes me think it may not be the best way to go forward, though I'm personally OK with the idea.

> But I don't know if you any strong objections?

I don't have strong objections. Perhaps we could rethink the name as you can use Combiner to copy a single (filtered) database to another so 'combiner' might have too narrow meaning.

**Comment by jkiviluo** [comment]
Another option is to use the 'transformer' that could just pass the DB as is (if there is no need to transform or filter). This is related to the issue that the transformer should get the capabilities of the combined importer/exporter too, so that we would have a way to directly change data structures (without going to e.g. csv in-between). Maybe there should be a new name that would better reflect all things the 'new transformer' could then do and that the user would also understand that it can be used to just pass data.

Pointing to same DB is a nice feature, but it is bit obscure and we shouldn't have central functionality relying on it (I think).

**Comment by manuelma** [comment]
> I don't have strong objections

Ok cool. But would you prefer to have the 'Combiner' or to just keep the current design with executable Data Stores?

> Perhaps we could rethink the name as you can use Combiner to copy a single (filtered) database to another

You're right. Do you have any ideas? English is hard...

> Another option is to use the 'transformer' that could just pass the DB as is (if there is no need to transform or filter). This is related to the issue that the transformer should get the capabilities of the combined importer/exporter too, so that we would have a way to directly change data structures (without going to e.g. csv in-between). Maybe there should be a new name that would better reflect all things the 'new transformer' could then do and that the user would also understand that it can be used to just pass data.

I think it's difficult when an item sometimes just passes data and sometimes writes. That leads to problems such as the one in the original post. In my view, it's better if we split the different actions into different items for better isolation. It is often the case that the user wants to execute just some of the actions and if they are lumped into the same item then it's impossible. Also, dividing the actions across different items makes it easier to loop to specific parts. What do you think? Maybe the Combiner could be the first version of the Exporter/Importer if we rename it, where the data is merged unaltered. And in the future, we could add the specification to it, so we could do more fancy things?

**Comment by soininen** [comment]
> But would you prefer to have the 'Combiner' or to just keep the current design with executable Data Stores?

Well, if you ask it that way, I'd prefer the current status quo. Then again, it would entail the duplicate Data stores as discussed above and those might be unintuitive for users. It makes me a bit torn here.

About possible names: 'Writer', 'Updater', 'Cloner', 'Scriber', 'Composer' come to my mind. Naming is hard indeed :)

**Comment by manuelma** [comment]
How about Transferrer?

**Comment by soininen** [comment]
> How about Transferrer?

I hope noone mixes that up with Transformer.

**Comment by manuelma** [comment]
How do we decide here? @ererkka ?

**Comment by DillonJ** [comment]
What about an "Alias" tool?

**Comment by manuelma** [comment]
I like aliaser as it captures the single DB import as well as the combination of multiple DBs.

**Comment by jkiviluo** [comment]
Doesn't "alias" mean that it's a pointer to the same data. But here it's about copying the data, right? If that's the case, I would prefer 'cloner' or even 'copier'.

Naturally we should think about future also. If the item does multiple things in the future, the name should already reflect that, otherwise we will need to rename again. What items will we have once Toolbox is ready? 
- Combiner (for combining more than one DB into one)
- Cloner (for copying all or part of a DB)
- Transformer (to alter data while also copying)
Or will some of these be combined? Combiner may be distinct functionality that is best left separate, but Cloner and Transformer could well fit in a single interface, me thinks.

**Comment by manuelma** [comment]
> But here it's about copying the data, right? 

It's about copying, but the main point is to combine several dbs into another db. Maybe the Combiner should only run if there is more than one input DB, otherwise it complains. Because there shouldn't be a point in making a copy of an existing db?

**Comment by DillonJ** [comment]
I think @soininen's suggestion here https://github.com/Spine-project/Spine-Toolbox/discussions/1624 merits some discussion in this context... we might need to step back a little in terms of design to try and find a nice solution here before rushing into anything

**Comment by manuelma** [comment]
I'm not a huge fan of #1624 but it requires the Merger anyways I believe.

**Comment by manuelma** [comment]
I run into a similar problem with executable Data Stores. In the workflow below, I want to load some template data from the Template DS into Data Store 1. After that, I want to run the Tool to write some additional data on top. The way it is now, Data Store 1 will always execute after the Tool. There is no way to achieve what I want in a single execution (I need to run two 'selection' executions). 

![Screenshot from 2021-12-13 08-20-02](https://user-images.githubusercontent.com/11472756/145769500-5b78e50d-a100-461c-9035-afb141118cbf.png)



If I had a Merger I could use a 'pure dependency link' to make the Merger always run before the Tool, like this:

![Screenshot from 2021-12-13 08-23-36](https://user-images.githubusercontent.com/11472756/145769452-537dabb9-4ddd-479b-9ecb-52d9b2521922.png)

I don't know if it's possible to do the same with Data Stores pointing to the same url.

Any thoughts?

**Comment by manuelma** [comment]
I feel we got stuck here. Any ideas how to proceed? I have the Combiner/Merger/Blender working in another branch, and I don't want it to get obsolete.

Maybe the decision to take is whether or not we want to keep executing Data Stores?

**Comment by DillonJ** [comment]
Myabe there's a higher level roadblock discussion to be had but for now it feels like if we have something that works we should go with that?

**Comment by soininen** [comment]
What about existing projects with DS-DS connections like the one in the issue descirption? How are we planning to handle backward compatiblity if DSes stop executing?

**Comment by manuelma** [comment]
> What about existing projects with DS-DS connections like the one in the issue descirption? How are we planning to handle backward compatiblity if DSes stop executing?

Project upgrader, I already did it when removing Combiner. We can reuse part of that code, basically we need to insert the necessary combiners in between DS-DS links.

**Comment by soininen** [comment]
I guess resurrecting Combiner is okay, then.

**Comment by manuelma** [comment]
Thank you! Is it Ok if I name it Merger? Another idea was Blender...

**Comment by soininen** [comment]
I vote for Merger.

**Comment by jkiviluo** [comment]
Merger is fine for me too. Maybe we need to think these more in the future, but at least Manuel can get this working now.

**Comment by DillonJ** [comment]
Mergificatinizeator ? 

More seriously, Merger sounds good.
